File: src/main/java/com/teamwizardry/librarianlib/gui/components/ComponentSlot.java
package com.teamwizardry.librarianlib.gui.components;

import com.teamwizardry.librarianlib.gui.GuiComponent;
import com.teamwizardry.librarianlib.gui.HandlerList;
import com.teamwizardry.librarianlib.gui.Option;
import com.teamwizardry.librarianlib.math.Vec2d;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.client.renderer.RenderItem;
import net.minecraft.item.ItemStack;
import net.minecraft.util.text.TextFormatting;

import java.util.List;

public class ComponentSlot extends GuiComponent<ComponentSlot> {
	
	public final Option<ComponentSlot, ItemStack> stack = new Option<>(null);
	public final Option<ComponentSlot, Boolean> tooltip = new Option<>(null);
	public final HandlerList<ISlotTextEventHandler<ComponentSlot>> quantityText = new HandlerList<>();
	public final HandlerList<ISlotInfoEventHandler<ComponentSlot>> itemInfo = new HandlerList<>();
	
	public ComponentSlot(int posX, int posY) {
		super(posX, posY, 16, 16);
	}

	@Override
	public void drawComponent(Vec2d mousePos, float partialTicks) {
        RenderHelper.enableGUIStandardItemLighting();
        GlStateManager.enableRescaleNormal();
		
		ItemStack stack = this.stack.getValue(this);
		String str = "" + stack.stackSize;
		str = quantityText.fireModifier(str, (h, v) -> h.handle(this, v));
		
		RenderItem itemRender = Minecraft.getMinecraft().getRenderItem();
        itemRender.zLevel = 200.0F;
        
        FontRenderer font = null;
        font = stack.getItem().getFontRenderer(stack);

        itemRender.renderItemAndEffectIntoGUI(stack, pos.xi, pos.yi);
        itemRender.renderItemOverlayIntoGUI(font == null ? Minecraft.getMinecraft().fontRendererObj : font, stack, pos.xi, pos.yi, str);
        
        itemRender.zLevel = 0.0F;
        
        
        if(mouseOverThisFrame && tooltip.getValue(this))
        	drawTooltip(stack, mousePos);
        
        GlStateManager.disableRescaleNormal();
        RenderHelper.disableStandardItemLighting();
	}
	
	public void drawTooltip(ItemStack stack, Vec2d mousePos) {
		List<String> list = stack.getTooltip(Minecraft.getMinecraft().thePlayer, Minecraft.getMinecraft().gameSettings.advancedItemTooltips);

        for (int i = 0; i < list.size(); ++i)
        {
            if (i == 0)
            {
                list.set(i, stack.getRarity().rarityColor + list.get(i));
            }
            else
            {
                list.set(i, TextFormatting.GRAY + list.get(i));
            }
        }

        itemInfo.fireAll((h) -> h.handle(this, list));
        
        FontRenderer font = stack.getItem().getFontRenderer(stack);
        setTooltip(list, font == null ? Minecraft.getMinecraft().fontRendererObj : font);
	}
	
	@FunctionalInterface
	public interface ISlotTextEventHandler<T> {
		String handle(T component, String text);
	}
	
	@FunctionalInterface
	public interface ISlotInfoEventHandler<T> {
		void handle(T component, List<String> info);
	}

}

File: src/main/java/com/teamwizardry/librarianlib/client/gui/components/ComponentSlot.kt
package com.teamwizardry.librarianlib.client.gui.components

import com.teamwizardry.librarianlib.client.gui.GuiComponent
import com.teamwizardry.librarianlib.client.gui.HandlerList
import com.teamwizardry.librarianlib.client.gui.Option
import com.teamwizardry.librarianlib.common.util.math.Vec2d
import com.teamwizardry.librarianlib.common.util.plus
import com.teamwizardry.librarianlib.common.util.stackSize
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.item.ItemStack
import net.minecraft.util.text.TextFormatting

open class ComponentSlot(posX: Int, posY: Int) : GuiComponent<ComponentSlot>(posX, posY, 16, 16) {

    val stack = Option<ComponentSlot, ItemStack?>(null)
    val tooltip = Option<ComponentSlot, Boolean>(true)
    val quantityText = HandlerList<(ComponentSlot, String?) -> String?>()
    val itemInfo = HandlerList<(ComponentSlot, MutableList<String>) -> Unit>()

    override fun drawComponent(mousePos: Vec2d, partialTicks: Float) {
        RenderHelper.enableGUIStandardItemLighting()
        GlStateManager.enableRescaleNormal()

        val stack = this.stack.getValue(this)
        if (stack != null) {
            var str = "" + stack.stackSize
            str = quantityText.fireModifier(str, { h, v -> h(this, v) }) ?: ""

            val itemRender = Minecraft.getMinecraft().renderItem
            itemRender.zLevel = 200.0f

            itemRender.renderItemAndEffectIntoGUI(stack, pos.xi, pos.yi)
            itemRender.renderItemOverlayIntoGUI(stack.item.getFontRenderer(stack) ?: Minecraft.getMinecraft().fontRendererObj, stack, pos.xi, pos.yi, str)

            itemRender.zLevel = 0.0f


            if (mouseOver && tooltip.getValue(this))
                drawTooltip(stack)
        }

        GlStateManager.disableRescaleNormal()
        RenderHelper.disableStandardItemLighting()
    }

    fun drawTooltip(stack: ItemStack) {
        val list = stack.getTooltip(Minecraft.getMinecraft().player, Minecraft.getMinecraft().gameSettings.advancedItemTooltips)

        for (i in list.indices) {
            if (i == 0) {
                list[i] = stack.rarity.rarityColor + list[i]
            } else {
                list[i] = TextFormatting.GRAY + list[i]
            }
        }

        itemInfo.fireAll { h -> h(this, list) }

        val font = stack.item.getFontRenderer(stack)
        setTooltip(list, font ?: Minecraft.getMinecraft().fontRendererObj)
    }

}
